nonoverlapping sampling

MathLens opened the pull request.

chenhongqi0716 committed:
add nonoverlapping arg

add nonoverlapping arg in R part and Bindings

MathLens commented on r-package/grf/bindings/QuantileForestBindings.cpp:
```diff
@@ -42,8 +44,18 @@ Rcpp::List quantile_train(std::vector<double> quantiles,
+                            honesty_fraction, honesty_prune_leaves, alpha, imbalance_penalty, num_threads, seed, clusters, samples_per_cluster,
+                            blocklength, blocknum};
+    if(nonoverlap){
+      options = ForestOptions{num_trees, ci_group_size, sample_fraction, mtry, min_node_size, honesty,
```
We do not need to create another new constructor for ForestOptions. Just add one more arg to the current one.

chenhongqi0716 replied:
> We do not need to create another new constructor for ForestOptions. Just add one more arg to the current one.

Done.

chenhongqi0716 committed:
add nonoverlap arg

chenhongqi0716 committed:
add nonoverlap arg into constructor

MathLens committed:
add arg

MathLens committed:
add interface

chenhongqi0716 committed:
Update RandomSampler.h

Just change the description of comment. Usually we use overlapping/nonoverlapping not repeating, because repeating/non repeating might be confused with thereplacement(nonreplacement) which we can add in the future,

chenhongqi0716 committed:
Update quantile_forest.R

A typo

MathLens committed:
implement

MathLens committed:
Merge branch 'overlap' of https://github.com/chenhongqi0716/grf into overlap

MathLens commented:
@chenhongqi0716 code finished. you can try it out before we do the final review.

chenhongqi0716 commented:
> @chenhongqi0716 code finished. you can try it out before we do the final review.

Thanks a lot! Let me try it tommorow.

chenhongqi0716 commented:
> @chenhongqi0716 code finished. you can try it out before we do the final review.

It works well. We can start the final review.

chenhongqi0716 committed:
Add explanations of nonoverlap

MathLens committed:
Update SamplingOptions.h

MathLens reviewed (commented):
Looks good to me

chenhongqi0716 commented on core/src/sampling/RandomSampler.cpp:
```diff
@@ -259,7 +260,22 @@ void RandomSampler::shuffle_and_split_block_no_overlap(std::vector<size_t>& samp
+  for (int i = 0; i < n_all - block_size + 1; ++i) {
+    heads.insert(i);
+  }
+  for (int i = 0; i < size; ++i) {
+    size_t head = random_number_generator() % heads.size();
+    for (int j = 0; j < block_size; ++j) {
+      // add the selected block
+      samples.push_back(head + j);
+      // delete heads causing overlap
+      heads.erase(head + j);
+      heads.erase(head - j);
+    }
+  }
```
I have a question here. I don't quite understand this part. If we delete the heads causing overlaps in each i loop, will it influence the number of blocks we want?
I mean since i starts from 0 to size-1, will it return in total 'size' number of blocks blocks even there are blocks overlapping (of cause, we delete those overlapping blocks) with each other?

MathLens replied:
no selected blocks are deleted. heads causing overlap are deleted so that overlapping blocks never appears in our selection.

chenhongqi0716 replied:
> no selected blocks are deleted. heads causing overlap are deleted so that overlapping blocks never appears in our selection.

Got it. Thanks!

chenhongqi0716 commented:
I suppose it's good now. So can we merge it?
